# Assigment 5


<a id='index-0'></a>

In this notebook we use the numpy library once again. As in assignment 2, we will also use the sympy library that we import. The sympy library namely contains the function rref which will return the reduced row echelon form of a matrix. 

In [1]:
import numpy as np
from sympy import *

## Linear combination and span of vectors

**Linear combination of vectors**

A linear combination of the vectors $\underline{w}_1, . . . , \underline{w}_n$ of length $m$ is any vector of the form $c_1\underline{w}_1 + c_2\underline{w}_2 + . . . c_n\underline{w}_n$ where $c_1, c_2, . . . , c_n$ are numbers.

Consider the vectors

$$\underline{w}_1=\begin{bmatrix}
    1\\
    1\\
    0 
\end{bmatrix}, \underline{w}_2=\begin{bmatrix}
    1\\
    0\\
    1 
\end{bmatrix}, \underline{w}_3=\begin{bmatrix}
    0\\
    1\\
    1 
\end{bmatrix}   $$

We write this in python as demonstrated below.

In [2]:
w1 = np.matrix([[1],[1],[0]])
w1

matrix([[1],
        [1],
        [0]])

In [3]:
w2 = np.matrix([[1],[0],[1]])
w2

matrix([[1],
        [0],
        [1]])

In [4]:
w3 = np.matrix([[0],[1],[1]])
w3

matrix([[0],
        [1],
        [1]])

We provide two examples of linear combination of $\underline{w}_1$, $\underline{w}_2$, $\underline{w}_3$. First, the vector

\begin{align}
\underline{a}&=2\underline{w}_1+ \underline{w}_2 + 3\underline{w}_3\\
& =2\begin{bmatrix}
    1\\
    1\\
    0 
\end{bmatrix}+\begin{bmatrix}
    1\\
    0\\
    1 
\end{bmatrix}+3\begin{bmatrix}
    0\\
    1\\
    1 
\end{bmatrix} \\
& = \begin{bmatrix}
    3\\
    5\\
    4 
\end{bmatrix}
\end{align}

is a linear combination of $\underline{w}_1$, $\underline{w}_2$, $\underline{w}_3$ with $c_1 = 2$, $c_2 = 1$ and $c_3 = 3$.

In [5]:
a=2*w1+w2+3*w3
a

matrix([[3],
        [5],
        [4]])

**Exercise**

Let $\underline{b}$ be a linear combination of $\underline{w}_1$, $\underline{w}_2$, $\underline{w}_3$ with $c_1 = -1$, $c_2 = 0$ and $c_3 = 2$. Find the vector $\underline{b}$.

In [6]:
# Your code here

**Note: for the following part we will use the sympy library, since we want to use the rref() function. That means that we enter vectors and matrices using Matrix() and not using np.matrix() as we did above.**

Consider the vectors:

$$\underline{w}_1=\begin{bmatrix}
    1\\
    1\\
    0 
\end{bmatrix}, \underline{w}_2=\begin{bmatrix}
    1\\
    0\\
    1 
\end{bmatrix}, \underline{b}=\begin{bmatrix}
    3\\
    2\\
    1 
\end{bmatrix}   $$

We write this in python as demonstrated below.

In [7]:
w1 = Matrix([[1],[1],[0]])
w1

Matrix([
[1],
[1],
[0]])

In [8]:
w2 = Matrix([[1],[0],[1]])
w2

Matrix([
[1],
[0],
[1]])

In [9]:
b = Matrix([[3],[2],[1]])
b

Matrix([
[3],
[2],
[1]])

We want to check whether $\underline{b}$ is a linear combination of $\underline{w}_1$ and $\underline{w}_2$. Hence, we want to find number $c_1$ and $c_2$ such that 

$$c_1\underline{w}_1 + c_2\underline{w}_2 = \underline{b}.$$

This is equivalent to 

$$c_1 \begin{bmatrix}
    1\\
    1\\
    0 
\end{bmatrix}+c_2\begin{bmatrix}
    1\\
    0\\
    1 
\end{bmatrix}=\begin{bmatrix}
    3\\
    2\\
    1 
\end{bmatrix} .$$

This we can write into

$$\begin{bmatrix}
    1 \ \ \ 1\\
    1 \ \ \ 0\\
    0 \ \ \ 1
\end{bmatrix}\begin{bmatrix}
    c_1\\
    c_2
\end{bmatrix}=\begin{bmatrix}
    3\\
    2\\
    1 
\end{bmatrix} .$$

Clearly, the equality given above boils down to a system of linear equations with extended matrix 

$$
\left[
\begin{array}{cc|c}
1 & 1 & 3 \\
1 & 0 & 2 \\
0 & 1 & 1\\
\end{array}
\right]
$$

In [10]:
Ab = Matrix([[1, 1, 3],[1, 0, 2],[0,1,1]])
Ab

Matrix([
[1, 1, 3],
[1, 0, 2],
[0, 1, 1]])

We apply the row-reduction procedure to solve this system of linear equations. The matrices according to the different steps of the procedure are separated by a ~

$$
\left[
\begin{array}{cc|c}
1 & 1 & 3 \\
1 & 0 & 2 \\
0 & 1 & 1\\
\end{array}
\right] \text{~} \left[
\begin{array}{cc|c}
1 & 1 & 3 \\
0 & -1 & -1 \\
0 & 1 & 1\\
\end{array}
\right] \text{~} \left[
\begin{array}{cc|c}
1 & 1 & 3 \\
0 & 1 & 1 \\
0 & 1 & 1\\
\end{array}
\right] \text{~} \left[
\begin{array}{cc|c}
1 & 0 & 2 \\
0 & 1 & 1 \\
0 & 0 & 0\\
\end{array}
\right]
$$

Recall that in python we can find the reduced form of the system of linear equations using rref(). This returns two elements. The first is the reduced row echelon form, and the second is the indices of the pivot columns. By typing Ab_rref[0], only the reduced row echelon form is printed.

In [11]:
Ab_rref = Ab.rref() 
Ab_rref[0]

Matrix([
[1, 0, 2],
[0, 1, 1],
[0, 0, 0]])

Hence, $c_1 = 2$ and $c_2 = 1$. We can conclude that $\underline{b} = 2\underline{w}_1 + 1\underline{w}_2$ and henceforth $\underline{b}$ is a linear combination of $\underline{w}_1$ and $\underline{w}_2$. 

**Exercise**

Consider the vectors:

$$\underline{w}_1=\begin{bmatrix}
    1\\
    1\\
    0 
\end{bmatrix}, \underline{w}_2=\begin{bmatrix}
    1\\
    0\\
    1 
\end{bmatrix}, \underline{b}=\begin{bmatrix}
    0\\
    1\\
    1 
\end{bmatrix} .$$

Check whether $\underline{b}$ is a linear combination of $\underline{w}_1$ and $\underline{w}_2$. Hence, find number $c_1$ and $c_2$ such that 

$$c_1\underline{w}_1 + c_2\underline{w}_2 = \underline{b}.$$

In [12]:
# Your code here

**Note: for the following part we will no longer use the sympy library. That means that we enter vectors and matrices using np.matrix() and not using Matrix().**

**Distance vector to a span of vectors**

Recall the following from the reader: Let $\underline{w}_1, . . . , \underline{w}_n$ and $\underline{b}$ be vectors of length $m$. Then the minimum distance of vector $\underline{b}$ to the $Span[\underline{w}_1, . . . , \underline{w}_n]$ is the length of the vector $\underline{b} − \underline{p}$ that satisfies the following two conditions:

(i) $\underline{p}=A\underline{c}$

(ii) $A^T(\underline{b}-\underline{p})=\underline{0}$,

Then the vector $\underline{c}$ is obtained by solving the normal equation

$$\underline{c}=(A^TA)^{-1}A^T\underline{b}.$$

Observe, as soon $\underline{c}$ is determined, the vector $\underline{p}$ is readily obtained using $\underline{p} = A\underline{c}$ from condition (i). So, the normal equation can be used to determine the distance of a vector $\underline{b}$ to a $Span[\underline{w}_1, . . . , \underline{w}_n]$. 

**Example**
Consider the vectors:

$$\underline{b}=\begin{bmatrix}
    -1\\
    -5\\
    10 
\end{bmatrix}, \underline{w}_1=\begin{bmatrix}
    5\\
    -2\\
    1 
\end{bmatrix}, \underline{w}_2=\begin{bmatrix}
    1\\
    2\\
    -1 
\end{bmatrix}   $$

We write this in python as demonstrated below.

In [13]:
b = np.matrix([[-1],[-5],[10]])
b

matrix([[-1],
        [-5],
        [10]])

In [14]:
w1 = np.matrix([[5],[-2],[1]])
w1

matrix([[ 5],
        [-2],
        [ 1]])

In [15]:
w2 = np.matrix([[1],[2],[-1]])
w2

matrix([[ 1],
        [ 2],
        [-1]])

The columns of the matrix $A$ consist of $\underline{w}_1$ and $\underline{w}_2$, respectively. Hence,

$$A =
\left[
\begin{array}{cc}
5 & 1 \\
-2 & 2 \\
1 & -1
\end{array}
\right].
$$

In [16]:
A = np.matrix([[5, 1],[-2, 2], [1, -1]])
A

matrix([[ 5,  1],
        [-2,  2],
        [ 1, -1]])

In python we can easily calculate $\underline{c}=(A^TA)^{-1}A^T\underline{b}$ using the function $A.I$ to calculate the inverse of matrix $A$ and using the function $np.transpose(A)$ to calculate $A^T$.

We first calculate $A^TA$:

In [17]:
ATA = np.transpose(A)*A
ATA

matrix([[30,  0],
        [ 0,  6]])

Next, we calculate $(A^TA)^{-1}$:

In [18]:
ATAinverse = ATA.I
ATAinverse

matrix([[0.03333333, 0.        ],
        [0.        , 0.16666667]])

Now we calculate $A^Tb$:

In [19]:
ATb = np.transpose(A)*b
ATb

matrix([[ 15],
        [-21]])

Finally, we calculate $c$:

In [20]:
c = ATAinverse*ATb
c

matrix([[ 0.5],
        [-3.5]])

Now, we can calculate $\underline{p}$ since

$$\underline{p}=A\underline{c}=\left[
\begin{array}{cc}
5 & 1 \\
-2 & 2 \\
1 & -1
\end{array}
\right]\left[
\begin{array}{c}
\frac{1}{2}\\
-\frac{7}{2}
\end{array}
\right]=\left[
\begin{array}{c}
-1\\
-8\\
4
\end{array}
\right]$$

In [21]:
p = A*c
p

matrix([[-1.],
        [-8.],
        [ 4.]])

Finally, we can calculate $|\underline{b}-\underline{p}|$. Recall from assignment 4 that we can use the function np.linalg.norm() to calculate this.

In [22]:
distance = np.linalg.norm(b-p)
distance

6.708203932499369

So, the distance of $\underline{b}$ to $Span[\underline{w}_1,\underline{w}_2]$ is $\sqrt{45}\approx 6.708$

**Exercise 4.8**

Determine the vector $\underline{c}$ in the normal equation $\underline{c}=(A^TA)^{-1}A^T\underline{b}$ where $\underline{b}= \begin{bmatrix}
    1\\
    2\\
    3 
\end{bmatrix}$ and $A =
\left[
\begin{array}{cc}
-2 & 2 \\
1 & 5 \\
1 & -1
\end{array}
\right].
$

In [23]:
# Your code here